app/models/volunteer.rb

```diff
@@ -113,35 +113,57 @@ class Volunteer < ApplicationRecord
       .merge(Assignment.inactive)
       .where.not(assignments: { volunteer_id: with_active_assignments.ids })
   }
+
   scope :will_take_more_assignments, (-> { where(take_more_assignments: true) })
-  scope :active, (-> { accepted.with_active_assignments })
 
-  scope :accepted_joined, (-> { accepted.left_outer_joins(:assignments) })
-  scope :loj_without_assignments, (-> { accepted_joined.where(assignments: { id: nil }) })
-  scope :loj_active_take_more, lambda {
-    accepted_joined.will_take_more_assignments.where(assignments: { id: Assignment.active.ids })
+  scope :activeness_not_ended, lambda {
+    where('activeness_might_end IS NULL OR activeness_might_end > ?', Time.zone.today)
+  }
+  scope :activeness_ended, lambda {
+    where(active: true)
+      .where('activeness_might_end IS NOT NULL AND activeness_might_end < ?', Time.zone.today)
+  }
+  scope :active, lambda {
+    activeness_not_ended.where(active: true)
   }
 
+  scope :inactive, lambda {
+    seeking_clients
+  }
   scope :seeking_clients, lambda {
-    accepted_joined
-      .merge(Assignment.inactive)
-      .where.not(assignments: { volunteer_id: with_active_assignments.ids })
-      .or(loj_without_assignments)
+    accepted.where(active: false).or(accepted.activeness_ended)
   }
   scope :seeking_clients_will_take_more, lambda {
-    accepted_joined
-      .merge(Assignment.inactive)
-      .where.not(assignments: { volunteer_id: with_active_assignments.ids })
-      .or(loj_without_assignments)
-      .or(loj_active_take_more)
+    seeking_clients.or(accepted.will_take_more_assignments)
   }
 
+  def verify_and_update_state
+    update(active: active?, activeness_might_end: relevant_period_end_max)
+  end
+
+  def relevant_period_end_max
+    # any assinment with no end means activeness is not going to end
+    return nil if group_assignments.stay_active.any? || assignments.stay_active.any?
+    group_assignment_max, assignment_max = period_end_maxes
+    if group_assignment_max.nil?
+      assignment_max
+    elsif assignment_max.nil?
+      group_assignment_max
+    else
+      [group_assignment_max, assignment_max].max # get the later date
+    end
+  end
+
+  def period_end_maxes
+    [group_assignments.active.maximum(:period_end), assignments.active.maximum(:period_end)]
+  end
+
   def active?
-    accepted? && (assignments.active.any? || group_assignments.ongoing.any? || group_assignments.no_end.any?)
+    accepted? && assignments.active.any? || group_assignments.active.any?
```

Comment:
Parenthesis is needed because `&&` has higher [priority](https://ruby-doc.org/core-2.2.0/doc/syntax/precedence_rdoc.html) than `||`.

Reply:
nope, your wrong. This is boolean...
```ruby
true && false || true  # returns true
true && true || false # returns true
true && true || true # returns true
true && false || false # returns false
```
Try it in your shell if you don't beleave me. This is simply how boolean logic works

Reply:
paranteses makes no difference, they are redundant. Boolean logic in Ruby works from left to right

Reply:
If we need the check to be done from left to right, the parenthesis not only is not needed, but also would be wrong to use. 
The reason that this is executed from left to right is that logic-boolean and has higher priority than logic-boolean or. That is not my personal opinion. It is suggested in the link provided and it is a rule in all logic operations. 
The reason is not that they have have the same priority. If they did, then, of course, again it would be executed from left to right. Operation of equal priority are indeed always executed from left to right.

Reply:
In any case `false && false || true ` will give true and I guess we wouldn't like a volunteer that is not accepted to be considered as active.

Reply:
damn, I hate to be wrong 🤦‍♂️ ... your right